### PR TITLE
Add CRUD products

### PR DESCRIPTION
**Pull request by maciejb2k**

### CODE REVIEW

File: app/controllers/api/v1/recipes_controller.rb [hunk 1]
```diff
@@ -34,12 +35,49 @@ def destroy
           render json: { errors: @recipe.errors.to_s }, status: :unprocessable_entity
         end
       end
+      
+      def createproduct
```
Comment (h4b00) [inline app/controllers/api/v1/recipes_controller.rb]:
```suggestion
      def add_product
```

File: app/controllers/api/v1/recipes_controller.rb [hunk 2]
```diff
@@ -34,12 +35,49 @@ def destroy
...
+        Product.find(params[:product_id])
+        Recipe.find(params[:id])
```
Comment (h4b00) [inline app/controllers/api/v1/recipes_controller.rb]: ?

File: app/controllers/api/v1/products_controller.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
+module Api
+  module V1
+    class ProductsController < ApplicationController
+      before_action :set_product, except: %i[create index]
+      
+      rescue_from ActiveRecord::RecordNotFound, with: :record_not_found
```
Comment (h4b00) [inline app/controllers/api/v1/products_controller.rb]: unnecessary

File: app/controllers/api/v1/recipes_controller.rb [hunk 4]
```diff
@@ -34,12 +35,49 @@ def destroy
...
+        
+        # Check if product arleady exists and return if present
+        if ProductRecipe.find_by(recipe_id: params[:id], product_id: params[:product_id]).present?  
+          render json: { errors: "Product arleady exists on recipe." }, status: :unprocessable_entity and return
+        end
```
Comment (h4b00) [inline app/controllers/api/v1/recipes_controller.rb]: replace with relation and add validation to migration & model

File: app/controllers/api/v1/recipes_controller.rb [hunk 5]
```diff
@@ -34,12 +35,49 @@ def destroy
...
+        
+        if product_recipe.save
+          render json: product_recipe
+        else
+          render json: { errors: product_recipe.errors.messages }, status: :unprocessable_entity
+        end
+      end
+      
+      def destroyproduct
+        product_recipe = ProductRecipe.find_by(recipe_id: params[:id], product_id: params[:product_id])
```
Comment (h4b00) [inline app/controllers/api/v1/recipes_controller.rb]: product_recipe = @recipe.products_recipes.find_by!(product_id: @product.id)

File: app/models/product_recipe.rb [hunk 6]
```diff
@@ -0,0 +1,6 @@
+class ProductRecipe < ApplicationRecord
+  validates :quantity, presence: true, numericality: { is_integer: true, greater_than: 0}
```
Comment (h4b00) [inline app/models/product_recipe.rb:2]: add uniq validations for [:product, :recipe]

File: config/routes.rb [hunk 7]
```diff
@@ -2,7 +2,14 @@
   # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
   namespace :api do
     namespace :v1 do
-      resources :recipes
+      resources :recipes do
+        member do
+          post 'products/:product_id', to: 'recipes#createproduct'
```
Comment (h4b00) [inline config/routes.rb]: post :add_product

File: config/routes.rb [hunk 8]
```diff
@@ -2,7 +2,14 @@
...
+          delete 'products/:product_id', to: 'recipes#destroyproduct'
```
Comment (h4b00) [inline config/routes.rb]: delete :remove_product

**Review by h4b00 — APPROVED** [review]

**Review by wkdz6 — APPROVED** [review]

**Pull request merged**